packages/library/src/rpc-request-coalescer.ts

```diff
@@ -0,0 +1,63 @@
+import { IRpcTransport } from '@solana/rpc-transport/dist/types/transports/transport-types';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import fastStableStringify from 'fast-stable-stringify';
+import { shouldCoalesceRequest } from './rpc-request-coalescer-gating';
+
+type CoalescedRequest = {
+    readonly abortController: AbortController;
+    numConsumers: number;
+    readonly responsePromise: Promise<unknown>;
+};
+
+export function getRpcTransportWithRequestCoalescing(transport: IRpcTransport): IRpcTransport {
+    const coalescedRequestsByHash: Record<string, CoalescedRequest> = {};
+    return async function makeCoalescedHttpRequest<TResponse>(
+        config: Parameters<IRpcTransport>[0]
+    ): Promise<TResponse> {
+        const { payload, signal } = config;
+        if (!shouldCoalesceRequest(payload)) {
+            return await transport(config);
+        }
+        const requestHash = fastStableStringify([payload.method, payload.params]);
+        if (coalescedRequestsByHash[requestHash] == null) {
+            const abortController = new AbortController();
+            coalescedRequestsByHash[requestHash] = {
+                abortController,
+                numConsumers: 0,
+                responsePromise: (async () => {
+                    try {
+                        return await transport<TResponse>({
+                            ...config,
+                            signal: abortController.signal,
+                        });
+                    } finally {
+                        delete coalescedRequestsByHash[requestHash];
+                    }
+                })(),
+            };
+        }
+        const coalescedRequest = coalescedRequestsByHash[requestHash];
+        coalescedRequest.numConsumers++;
+        if (signal) {
+            const handleAbort = (e: AbortSignalEventMap['abort']) => {
+                signal.removeEventListener('abort', handleAbort);
+                coalescedRequest.numConsumers -= 1;
+                if (coalescedRequest.numConsumers === 0) {
+                    const { reason } = e.target as AbortSignal;
+                    const abortController = coalescedRequest.abortController;
+                    delete coalescedRequestsByHash[requestHash];
+                    abortController.abort(reason);
+                }
```

Comment:
Even if you trigger the `AbortSignal` on your request, don't nuke the fetch until you're the _last_ interested party.